### DISCUSSION TITLE
More than one guess control in Krotov

### DISCUSSION THREAD

**Discussion by STaherpour** [qucontrol/krotov#110] [Q&A] [answered]
Hi 
I have two guess controls in my example. How can I separate pulses from each guess controls during the optimization?
I mean we have all pulses in `opt_result.all_pulses` But how can we realize that which one is for control1 and which is for the other?

**Comment by goerz** [comment, marked as answer]
The `opt_result.all_pulses` are in the same order as the list that you get with [`extract_controls(objectives)`](https://qucontrol.github.io/krotov/v1.2.1/API/krotov.conversions.html#krotov.conversions.extract_controls). This is simply the order in which the controls occur in the generators of the different objectives. You could use [`extract_controls_mapping`](https://qucontrol.github.io/krotov/v1.2.1/API/krotov.conversions.html#krotov.conversions.extract_controls_mapping) to double check this, but the order should be quite intuitive.

**Reply by STaherpour** [reply]
Thanks.
I used `krotov.conversions.extract_controls_mapping(objectives, controls)` and I received `[[[[1], [2]]]]`
Does it mean, for example in 10 iterations of optimization the first array of all_pulses is about first control and the second array relates to second control and again the third array relates to first control?

**Reply by goerz** [reply]
I might have misunderstood your question;  `opt_result.all_pulses` is a list of lists of arrays. As the [documentation](https://qucontrol.github.io/krotov/v1.2.1/API/krotov.result.html#krotov.result.Result.all_pulses) says, **for each iteration**, it's a list of the optimized pulses (in the order corresponding to [guess_controls](https://qucontrol.github.io/krotov/v1.2.1/API/krotov.result.html#krotov.result.Result.guess_controls)).

So the first list in `all_pulses` relates to the first iteration of the optimization, and then the first array inside of that relates to the first control, the second to the second control, and so forth. I think you're missing that you get the pulses *from all iterations*.

`krotov.conversions.extract_controls_mapping(objectives, controls)` giving you `[[[[1], [2]]]]` just tells you that the first optimized pulse relates to the first control and the second optimized pulse relates to the second control, which is always going to be the case.

**Reply by STaherpour** [reply]
Thanks a lot.
